Sources/NIOConcurrencyHelpers/atomics.swift

```diff
@@ -458,11 +505,36 @@ public final class AtomicBox<T: AnyObject> {
     /// more than that this operation is atomic: there is no guarantee that any other
     /// event will be ordered before or after this one.
     ///
+    /// - warning: The implementation of `exchange` contains a _Compare and Exchange loop_, ie. it may busy wait with
+    ///            100% CPU load.
+    ///
     /// - Returns: The value of this object
     public func load() -> T {
-        let ptrBits = self.storage.load()
+        // step 1: We need to gain ownership of the value by successfully swapping 0 (marker value) in.
+        var ptrBits: UInt = 0
+        while true {
+            let speculativeVal = self.storage.load()
+            guard speculativeVal != 0 else {
+                sys_pthread_yield()
+                continue
+            }
+            if self.storage.compareAndExchange(expected: speculativeVal, desired: 0) {
+                ptrBits = speculativeVal
+                break
+            }
+        }
+
+        // step 2: We now consumed a +1'd version of val, so we have all the time in the world to retain it.
         let ptr = Unmanaged<T>.fromOpaque(UnsafeRawPointer(bitPattern: ptrBits)!)
-        return ptr.takeUnretainedValue()
+        let value = ptr.takeUnretainedValue()
+
+        // step 3: Now, let's exchange it back into the store
+        if !self.storage.compareAndExchange(expected: 0, desired: ptrBits) {
```

Comment:
I think we have an ABA problem here.

T1: load value A (set 0)
T2: store value B (overwrite 0)
T3: load value B (set 0)
T1: write value A back (works because value is 0)
T3: fails to write back value B, because T1 wrote value A back

we just lost the write of value B

Reply:
@drexin great catch, you're absolutely right! I thought I'd get away with not doing a CAS loop in `store` but no :).

Reply:
@drexin fixed (and tested in `testMultipleLoadsRacingWhilstStoresAreGoingOn`)